lib/bearerstrategy.js

```diff
@@ -118,121 +125,298 @@ const jws = require('jws');
+    log.info('We did not pass Req back to Callback');
+    return self._verify(verifiedToken, done);
   });
+}
 
-  function jwtVerify(req, token, done) {
-    const decoded = jws.decode(token);
-    let PEMkey = null;
+Strategy.prototype.authenticate = function authenticateStrategy(req) {
+  const self = this;
 
-    if (decoded == null) {
-      done(null, false, 'Invalid JWT token.');
-    }
+  /* Some introduction to async.waterfall (from the following link):
+   * http://stackoverflow.com/questions/28908180/what-is-a-simple-implementation-of-async-waterfall
+   *
+   *   Runs the tasks array of functions in series, each passing their results 
+   * to the next in the array. However, if any of the tasks pass an error to 
+   * their own callback, the next function is not executed, and the main callback
+   * is immediately called with the error.
+   *
+   * Example:
+   *
+   * async.waterfall([
+   *   function(callback) {
+   *     callback(null, 'one', 'two');
+   *   },
+   *   function(arg1, arg2, callback) {
+   *     // arg1 now equals 'one' and arg2 now equals 'two'
+   *     callback(null, 'three');
+   *   },
+   *   function(arg1, callback) {
+   *     // arg1 now equals 'three'
+   *     callback(null, 'done');
+   *   }
+   * ], function (err, result) {
+   *      // result now equals 'done'    
+   * }); 
+   */
+  async.waterfall([
 
-    log.info('token decoded:  ', decoded);
+    // compute metadata url
+    (next) => {
+      if (!self._options.certificate && !self._options.identityMetadata) {
+        log.warn('No options was presented to Strategy as required.');
+        return next(new TypeError(`BearerStrategy requires either a PEM encoded public key\ 
+          or a metadata location that contains cert data for RSA and ECDSA callback.`));
+      }
 
-    // We have two different types of token signatures we have to validate here. One provides x5t and the other a kid.
-    // We need to call the right one.
+      // (1) no metadata but have a certificate, pass `null` to the next function, 
+      // which will skip the metadata loading
+      if (!self._options.identityMetadata)
+        return next(null, null);
 
-    if (decoded.header.x5t) {
-      PEMkey = this.metadata.generateOidcPEM(decoded.header.x5t);
-    } else if (decoded.header.kid) {
-      PEMkey = this.metadata.generateOidcPEM(decoded.header.kid);
-    } else {
-      throw new TypeError('We did not reveive a token we know how to validate');
-    }
+      // (2) there is only metadata, we calulate the metadataURL and pass it to the
+      // next function for metadata loading
 
-    jwt.verify(token, PEMkey, options, (err, verifiedToken) => {
-      if (err) {
-        if (err instanceof jwt.TokenExpiredError) {
-          log.warn('Access token expired');
-          return done(null, false, 'The access token expired');
+      // default key for metadata cache
+      var cacheKey = 'ordinary';
+
+      var metadatURL = self._options.identityMetadata;
```

Comment:
suggest changing variable name from _options.identityMetadata to: _options.metadataUrl.